### PR TITLE
Feature/8796 create site ui

### LINKED ISSUE #8796: [Site Creation] Create site logic + ui
Send all the gather information to the server so it can create the site - use NewCreateSiteService

- block the UI (back button) when the site is being created

- show a notification if the user leaves the app while the site is being created
 -> 1) "We're creating your new site"
-> 2) "Your site has been created!"

- add retry button when the creation in the NewCreateSiteService fails

Use NewSiteCreationPreviewFragment for the UI + add a new ViewModel.

### PR DESCRIPTION
**Pull request by malinajirka**
Fixes #8796 

**NOTE: `feature/8512-create-site-step` is set as target branch. Replace it with `feature/master-site-creation` when #8795 is merged.**

This PR integrates NewSiteCreationService with the NewSitePreviewFragment + introduces `NewSitePreviewViewModel`

To test:
- My Site -> Switch Site -> Plus sign Icon -> Create a WordPress.com Site -> Select a Segment -> Skip Verticals -> Skip Site Info Screen -> type something 

Verify following + try to rotate your device

- turn on the airplane mode -> pick a domain -> no connection error is displayed ->  turn off the airplane mode -> click on retry -> the flow should continue

- pick a domain ->  when the loading indicator is displayed and you leave the app a progress notification is shown + it's automatically dismissed when you reenter the app

- pick a domain ->  when the loading indicator is displayed and you leave the app a progress notification is shown + if you wait until the site is created, the progress notification is replaced with 
"Site Created" notification + Tapping on the notification should redirect you to the app

- pick a domain -> wait until the site is created -> notice the Preview screen is displayed with an already preloaded site in the webview

- pick a domain -> wait until the preview screen is displayed -> click on the "OK" button -> notice you are redirected to the My Site screen and the newly created site is selected

- edit `CreateSiteUseCase` so it fails -> pick a domain -> generic error is shown

- verify the "Cancel/'X'" button on both error screens terminates the flow

Known issues
1) WebView's content is cleared on rotation #8821
2) When the user presses "OK" button and the SiteModel of the newly created site hasn't been fetched from the server -> the user should be navigated to the SitePickerActivity and a toast message should be displayed. #8822
3) The transition animation is missing #8583 
4) There is no timeout set when the site is being preloaded into the webview -> so if it takes 30seconds before the site is fully loaded, the user will see the loading screen for 30s #8824

creating

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchSiteUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    private var continuation: Continuation<OnSiteChanged>? = null
+
+    suspend fun fetchSite(siteId: Long, isWpCom: Boolean): OnSiteChanged {
+        return suspendCoroutine { cont ->
+            continuation = cont
+            dispatcher.dispatch(SiteActionBuilder.newFetchSiteAction(createWpComSiteModel(siteId, isWpCom)))
+        }
+    }
+
+    private fun createWpComSiteModel(siteId: Long, isWpCom: Boolean): SiteModel {
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchSiteUseCase.kt]: Oh, I've added the parameter and I forgot to rename the method. Nevertheless, the `FetchSiteUseCase` renamed to `FetchWpComSiteUseCase` in [a2a7c56](https://github.com/wordpress-mobile/WordPress-Android/pull/8825/commits/a2a7c5625995e653f781ea49553d0ae7e897321b)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchSiteUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import org.wordpress.android.fluxc.store.SiteStore
+import org.wordpress.android.fluxc.store.SiteStore.OnSiteChanged
+import javax.inject.Inject
+import kotlin.coroutines.experimental.Continuation
+import kotlin.coroutines.experimental.suspendCoroutine
+
+/**
+ * Transforms OnWPComSiteFetched EventBus event to a coroutine.
+ */
+class FetchSiteUseCase @Inject constructor(
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchSiteUseCase.kt]: Good point - the `FetchSiteUseCase` renamed to `FetchWpComSiteUseCase` in

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+                        throw IllegalStateException("Site successfully fetched but has not been found in the local db.")
+                    }
+                    return@launch
+                }
+                delay((attemptNumber + 1) * FETCH_SITE_BASE_RETRY_DELAY_IN_MILLIS) // +1 -> starts from 0
+            }
+        }
+    }
+
+    private fun startPreloadingWebView() {
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt]: Fixed in [a2a7c56](https://github.com/wordpress-mobile/WordPress-Android/pull/8825/commits/a2a7c5625995e653f781ea49553d0ae7e897321b)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        _onHelpClicked.call()
+    }
+
+    fun onCancelWizardClicked() {
+        _onCancelWizardClicked.call()
+    }
+
+    fun onOkButtonClicked() {
+        /**
+         * newlyCreatedSiteLocalId might be null if the fetchSite request failed or haven't finished yet.
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt]: I've decided to create a nonnullable state instead. I find it a bit more readable.
[e71e26d](https://github.com/wordpress-mobile/WordPress-Android/pull/8825/commits/e71e26dcdc5b8751daf96ce3c15f929c3f41e15d)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    fun start(siteCreationState: SiteCreationState) {
+        if (isStarted) {
+            return
+        }
+        isStarted = true
+        this.siteCreationState = siteCreationState
+        requireNotNull(siteCreationState.domain)
+        urlWithoutScheme = siteCreationState.domain!!
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt]: I keep forgetting about this:D. Thanks
Fixed in [407e1a5](https://github.com/wordpress-mobile/WordPress-Android/pull/8825/commits/407e1a5932b874326117d59c4594667939662b03)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+    /**
+     * Fetch newly created site model - supports retry with linear backoff.
+     */
+    private fun fetchNewlyCreatedSiteModel(remoteSiteId: Long) {
+        launch {
+            repeat(FETCH_SITE_NUMBER_OF_RETRIES) { attemptNumber ->
+                val onSiteFetched = fetchSiteUseCase.fetchSite(remoteSiteId, isWpCom = true)
+                if (!onSiteFetched.isError) {
+                    val siteBySiteId = siteStore.getSiteBySiteId(remoteSiteId)
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt]: Yep, definitely better. Thanks
[407e1a5](https://github.com/wordpress-mobile/WordPress-Android/pull/8825/commits/407e1a5932b874326117d59c4594667939662b03)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationPreviewFragment.kt [hunk 7]
```diff
@@ -8,34 +12,195 @@ import android.support.v7.app.AppCompatActivity
...
+        if (context !is OnHelpClickedListener) {
+            throw IllegalStateException("Parent activity must implement OnHelpClickedListener.")
+        }
+        sitePreviewScreenListener = context
+        helpClickedListener = context
+    }
+
+    override fun onResume() {
+        super.onResume()
+        mServiceEventConnection = ServiceEventConnection(activity!!, NewSiteCreationService::class.java, viewModel)
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationPreviewFragment.kt]: Is there any difference?
Fixed in [109753d](https://github.com/wordpress-mobile/WordPress-Android/pull/8825/commits/109753ddbfa7144679f452aae4da4fd540b468aa)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationPreviewFragment.kt [hunk 8]
```diff
@@ -8,34 +12,195 @@ import android.support.v7.app.AppCompatActivity
...
-    private val domainSpan: Pair<Int, Int> = Pair(Math.min(subdomainSpan.second, urlShort.length), urlShort.length)
+private const val ARG_DATA = "arg_site_creation_data"
+
+class NewSiteCreationPreviewFragment : NewSiteCreationBaseFormFragment<NewSiteCreationListener>(),
+        PageFullyLoadedListener {
+    /**
+     * We need to connect to the service, so the service knows when the app is in the background. The service
+     * automatically shows system notifications when site creation is in progress and the app is in the background.
+     */
+    private var mServiceEventConnection: ServiceEventConnection? = null
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationPreviewFragment.kt]: Fixed in [109753d](https://github.com/wordpress-mobile/WordPress-Android/pull/8825/commits/109753ddbfa7144679f452aae4da4fd540b468aa)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+            }
+        }
+    }
+
+    /**
+     * Fetch newly created site model - supports retry with linear backoff.
+     */
+    private fun fetchNewlyCreatedSiteModel(remoteSiteId: Long) {
+        launch {
+            repeat(FETCH_SITE_NUMBER_OF_RETRIES) { attemptNumber ->
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt]: Good call - fixed in [7e2f45d](https://github.com/wordpress-mobile/WordPress-Android/pull/8825/commits/7e2f45d6f21d83eeb185f19182fd25e1703a0b67)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt]: Let me know what you think - I can revert it:)

**Comment by malinajirka** [comment]
Thanks @oguzkocer! I've fixed all the issues + implemented the improvements. It's ready for another round.

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchWpComSiteUseCase.kt [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    private suspend fun fetchSite(siteId: Long): OnSiteChanged {
+        return suspendCoroutine { cont ->
+            continuation = cont
+            dispatcher.dispatch(SiteActionBuilder.newFetchSiteAction(createSiteModel(siteId)))
+        }
+    }
+
+    private fun createSiteModel(siteId: Long): SiteModel {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchWpComSiteUseCase.kt]: I think after the change in a2a7c5625995e653f781ea49553d0ae7e897321b, the original `createWpComSiteModel` is a better name here. Sorry if this wasn't clear :(

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSitePreviewViewModel.kt]: I love this change, thank you!

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchWpComSiteUseCase.kt [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * Transforms FETCH_SITE -> UPDATE_SITE fluxC request-response pair to a coroutine.
+ */
+class FetchWpComSiteUseCase @Inject constructor(
+    private val dispatcher: Dispatcher,
+    @Suppress("unused") private val siteStore: SiteStore
+) {
+    private var continuation: Continuation<OnSiteChanged>? = null
+
+    suspend fun fetchSite(remoteSiteId: Long, numberOfRetries: Int): OnSiteChanged {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchWpComSiteUseCase.kt]: How do you feel about adding a default value for the `numberOfRetries`? It'd be good to make that a `const` if we add it.

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchWpComSiteUseCase.kt (hunk 10 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchWpComSiteUseCase.kt]: I would have sweared I renamed it back... interesting

Fixed in [5b606c7](https://github.com/wordpress-mobile/WordPress-Android/pull/8825/commits/5b606c73b72526192808d5d0c9e0ba3cb174edba)

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchWpComSiteUseCase.kt (hunk 11 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/creation/FetchWpComSiteUseCase.kt]: I've add a default value and renamed the method in [e953b06](https://github.com/wordpress-mobile/WordPress-Android/pull/8825/commits/e953b06e90b7f3e1659bb5762fb29de874d940d4)

**Comment by malinajirka** [comment]
I've fixed the last two remaining issues. Thanks @oguzkocer for all the great suggestions for improvement.

File: WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationPreviewFragment.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationPreviewFragment.kt]: I went form not-using it all to overusing it. I need to find the right path.

**Pull request merged**